cores/esp8266/uart.h

```diff
@@ -110,43 +110,52 @@ extern "C" {
 
 #define UART_TX_FIFO_SIZE 0x80
 
-struct uart_;
-typedef struct uart_ uart_t;
+    typedef void(*uartInterruptHandler)(void* arg); // ICACHE_RAM_ATTR required
 
-uart_t* uart_init(int uart_nr, int baudrate, int config, int mode, int tx_pin, size_t rx_size);
-void uart_uninit(uart_t* uart);
+    struct uart_;
+    typedef struct uart_ uart_t;
 
-void uart_swap(uart_t* uart, int tx_pin);
-void uart_set_tx(uart_t* uart, int tx_pin);
-void uart_set_pins(uart_t* uart, int tx, int rx);
-bool uart_tx_enabled(uart_t* uart);
-bool uart_rx_enabled(uart_t* uart);
+    uart_t* uart_init(int uart_nr, int baudrate, int config, int mode, int tx_pin, size_t rx_size);
+    void uart_uninit(uart_t* uart);
 
-void uart_set_baudrate(uart_t* uart, int baud_rate);
-int uart_get_baudrate(uart_t* uart);
+    void uart_swap(uart_t* uart, int tx_pin);
+    void uart_set_tx(uart_t* uart, int tx_pin);
+    void uart_set_pins(uart_t* uart, int tx, int rx);
+    bool uart_tx_enabled(uart_t* uart);
+    bool uart_rx_enabled(uart_t* uart);
 
-size_t uart_resize_rx_buffer(uart_t* uart, size_t new_size);
-size_t uart_get_rx_buffer_size(uart_t* uart);
+    void uart_set_baudrate(uart_t* uart, int baud_rate);
+    int uart_get_baudrate(uart_t* uart);
 
-size_t uart_write_char(uart_t* uart, char c);
-size_t uart_write(uart_t* uart, const char* buf, size_t size);
-int uart_read_char(uart_t* uart);
-int uart_peek_char(uart_t* uart);
-size_t uart_read(uart_t* uart, char* buffer, size_t size);
-size_t uart_rx_available(uart_t* uart);
-size_t uart_tx_free(uart_t* uart);
-void uart_wait_tx_empty(uart_t* uart);
-void uart_flush(uart_t* uart);
+    size_t uart_resize_rx_buffer(uart_t* uart, size_t new_size);
+    size_t uart_get_rx_buffer_size(uart_t* uart);
 
-bool uart_has_overrun (uart_t* uart); // returns then clear overrun flag
-bool uart_has_rx_error (uart_t* uart); // returns then clear rxerror flag
+    size_t uart_write_char(uart_t* uart, char c);
+    size_t uart_write(uart_t* uart, const char* buf, size_t size);
+    int uart_read_char(uart_t* uart);
+    int uart_peek_char(uart_t* uart);
+    size_t uart_read(uart_t* uart, char* buffer, size_t size);
+    size_t uart_rx_available(uart_t* uart);
+    size_t uart_tx_free(uart_t* uart);
+    void uart_wait_tx_empty(uart_t* uart);
+    void uart_flush(uart_t* uart);
 
-void uart_set_debug(int uart_nr);
-int uart_get_debug();
+    bool uart_has_overrun(uart_t* uart); // returns then clear overrun flag
+    bool uart_has_rx_error(uart_t* uart); // returns then clear rxerror flag
 
-void uart_start_detect_baudrate(int uart_nr);
-int uart_detect_baudrate(int uart_nr);
+    void uart_set_debug(int uart_nr);
+    int uart_get_debug();
 
+    void uart_start_detect_baudrate(int uart_nr);
+    int uart_detect_baudrate(int uart_nr);
+
+    // uart_subscribeInterrupt & uart_unsubscribeInterrupt are not safe and must
+    // be called within ETS_UART_INTR_DISABLE()/ETS_UART_INTR_ENABLE() protection
+    //
```

Comment:
Other unsafe functions have `_unsafe_` part of their name.
I think that is a good reminder for their users.
Or, why not calling INTR_EN/DISABLE() from inside these functions and remove that warning ?

Reply:
- Rename them _unsafe, add new ones calling the unsafe version in a INTR_EN/DISABLE sandwhich

- too bad it's C, otherwise a parameter `bool protect` with a default value would do the job

- keep as it is, and let's rely on users' intelligence for reading doc and comments

Reply:
Please implement like bullet 1 above. The unsafe word in the function name is meant to indicate it's not interrupt-safe and it shouldn't be exported, if at all possible.
The unsafe functions are meant for resuse from other unsafe functions and/or for sandwich wrapping into safe versions.

Reply:
This is not quite resolved. I'll try to explain again.
I said:
> The unsafe word in the function name is meant to indicate it's not interrupt-safe and it shouldn't be exported

In this case, there could be undefined behavior if there is an IRQ in the middle of calling these functions.
Instead, mark the current implementations of these functions as static, and implement new functions without the _unsafe() word. Then export those in the .h. These exported versions call the _unsafe() functions inside a sandwich.

Reply:
Also, it is ok to skip argument sanity checks in the _unsafe functions, but the safe versions should implement checks on the received args.

Reply:
@devyte Maybe I was not clear enough.  You can't wrap the subscribe as the site of where you call it has to be already wrapped with it due to several processes of setting up the uart hardware are required to be called within the same DISABLE/ENABLE.  This is what I was pointing out by referencing the call spots withing UART.C.  This same setup; but different based on needs of the caller is done externally by anyone needing to use uart ISR.
IF DISABLE/ENABLE were actually ref-count like entities, then it would be a good practice, as the ENABLE inside would only ref-count and not truly enable; but this is not the case.

Reply:
YES, these two unsafe functions are called by the user.  I pointed to my project that consumes these in the pull comments at the top.

Can I just point them to my code?  Ok, its not simple ;-)

The real use of this is where you aren't doing simple serial, otherwise you would just use Serial.  In general, using the hardware to talk to devices that take a serial stream; often where translation is done from data on the MCU as you send.  Which means configuring the UART to your custom needs (using all those esoteric features to invert, turn off stop bits, etc) that are not exposed by Serial or even the current uart.h.

Reply:
Here is a link to my code that does this specifically
https://github.com/Makuna/NeoPixelBus/blob/959cfac07b97e95e10fc71f311f81fa3e2757895/src/internal/NeoEsp8266UartMethod.cpp#L75